documentation/docs/explanations/concurrency.md

```diff
@@ -0,0 +1,80 @@
+
+## The Actor Model
+
+Stronghold employs a well known concurrency architecture: the actor model. The basic idea of the actor model is to have isolated actors, each taking care of some functionality. Actors receive messages with data to act upon and return data when they finish processing it. Since each actor contains its own state and concurrency is achieved by not directly calling functions but by polling messages, most of the undesirable concurrency problems are taken care of. Dead-locks will never occur. 
+
+The actor system is almost ubiquitous in the target architecture. Many modern languages have the actor model built in with an excellent supervisor, among other tools. With Rust, the integration involves a lot of boilerplate code. But it’s for each user to decide whether or not the actor model approach is favorable. We wanted to know whether we could provide a simple interface, ideally some primitive types to work on with simple function calls, but still run in a concurrent setup without the headaches that come with locks and mutexes. 
+
+### Why didn't we keep the actor model?
+
+First iteration of Stronghold with the actor model was done with the rust crate called [Riker](https://riker.rs/) actor system crate. 
+Unfortunately this crate was discontinued we had to find an alternative which was the [Actix](https://github.com/actix/actix) actor framework crate.
+Actix uses the Tokio runtime which came as an issue for us as the Stronghold library. 
+Indeed software using Stronghold had to pass ownership of the runtime to our code when making function calls to the Stronghold library.
+This was cumbersome for our users and we abandoned the idea of using the actor model for concurrent programming and explored other paradigms.
+
+
+## Software Transactional Memory (STM) with Transactional Locking 2 (TL2)
+
+A newer approach is using Software Transactional Memory (STM). STMs have been around for quite some time. In STMs, each operation on memory happens in an atomic transaction. Whenever memory is modified, this modification is written into a log. While inside a transaction, reading from memory also is done through a log. The transaction has finished when all changes recorded inside the log have been committed to the actual memory. A transaction fails if another thread tries to modify the targeted piece of memory between operations. A failed transaction can be re-run any number of times. 
+
+This approach guarantees that modifications to memory are always consistent, but it comes with a restriction. Since transactions can be retried, operations inside a transaction must be idempotent and should not have any side effects. In an extreme case, think of a function that launches an ICMB: you can not reverse the process. Another edge case concerning STM-based approaches is interleaving transactions, where reads and writes are alternating between two threads. In a worst-case scenario, both transactions would retry indefinitely. 
```

Comment:
```suggestion
This approach guarantees that modifications to memory are always consistent, but it comes with a restriction. Since transactions can be retried, operations inside a transaction must be idempotent and should not have any side effects. In an extreme case, think of a function that launches an ICMB: you can not reverse the process. Another edge case concerning STM-based approaches is interleaving transactions, where reads and writes are alternating between two threads. In a worst-case scenario, both transactions would retry indefinitely.
```